core/src/main/scala/kafka/controller/KafkaControllerZkUtils.scala

```diff
@@ -0,0 +1,533 @@
+  (Map[String, LogConfig], Map[String, Exception]) = {
+    val logConfigs = mutable.Map.empty[String, LogConfig]
+    val failed = mutable.Map.empty[String, Exception]
+    val configResponses = try {
+      getTopicConfigs(topics)
+    } catch {
+      case e: Exception =>
+        topics.foreach(topic => failed.put(topic, e))
+        return (logConfigs.toMap, failed.toMap)
+    }
+    configResponses.foreach { configResponse =>
+      val topic = configResponse.ctx.asInstanceOf[String]
+      if (Code.get(configResponse.rc) == Code.OK) {
+        val overrides = ConfigEntityZNode.decode(configResponse.data)
+        val logConfig = LogConfig.fromProps(config, overrides.getOrElse(new Properties))
+        logConfigs.put(topic, logConfig)
+      } else if (Code.get(configResponse.rc) == Code.NONODE) {
+        val logConfig = LogConfig.fromProps(config, new Properties)
+        logConfigs.put(topic, logConfig)
+      } else {
+        failed.put(topic, KeeperException.create(Code.get(configResponse.rc)))
+      }
+    }
+    (logConfigs.toMap, failed.toMap)
+  }
+
+  def getAllBrokersInCluster: Seq[Broker] = {
+    val getChildrenResponse = retryUntilConnected(GetChildrenRequest(BrokerIdsZNode.path, null)).asInstanceOf[GetChildrenResponse]
+    if (Code.get(getChildrenResponse.rc) == Code.OK) {
+      val brokerIds = getChildrenResponse.children.map(_.toInt)
+      val getDataRequests = brokerIds.map(brokerId => GetDataRequest(BrokerIdZNode.path(brokerId), brokerId))
+      val getDataResponses = retryUntilConnected(getDataRequests).map(_.asInstanceOf[GetDataResponse])
+      getDataResponses.flatMap { getDataResponse =>
+        val brokerId = getDataResponse.ctx.asInstanceOf[Int]
+        if (Code.get(getDataResponse.rc) == Code.OK) {
+          Option(BrokerIdZNode.decode(brokerId, getDataResponse.data))
+        } else if (Code.get(getDataResponse.rc) == Code.NONODE) {
+          None
+        } else {
+          throw KeeperException.create(Code.get(getDataResponse.rc))
+        }
+      }
+    } else if (Code.get(getChildrenResponse.rc) == Code.NONODE) {
+      Seq.empty
+    } else {
+      throw KeeperException.create(Code.get(getChildrenResponse.rc))
+    }
+  }
+
+  def getAllTopicsInCluster: Seq[String] = {
+    val getChildrenResponse = retryUntilConnected(GetChildrenRequest(TopicsZNode.path, null)).asInstanceOf[GetChildrenResponse]
+    if (Code.get(getChildrenResponse.rc) == Code.OK) {
+      getChildrenResponse.children
+    } else if (Code.get(getChildrenResponse.rc) == Code.NONODE) {
+      Seq.empty
+    } else {
+      throw KeeperException.create(Code.get(getChildrenResponse.rc))
+    }
+  }
+
+  def setTopicAssignmentRaw(topic: String, assignment: Map[TopicAndPartition, Seq[Int]]) = {
+    val setDataRequest = SetDataRequest(TopicZNode.path(topic), TopicZNode.encode(assignment), -1, null)
+    retryUntilConnected(setDataRequest).asInstanceOf[SetDataResponse]
+  }
+
+  def getAllLogDirEventNotifications = {
```

Comment:
Could we explicitly define the return type of this method?

Reply:
Done.